### PR TITLE
docker-compose multi-stage build

### LINKED ISSUE #10: Move to multi-stage build for container-features
This issue proposes moving the container-features build over to using a multi-stage docker build.

Prior to container-features, it was possible to build the dev container image with an inline cache manifest, push it to a registry and then reference that image in the `cacheFrom` in `devcontainer.json` (or in `cache_from` in `docker-compose.yml`) to have the registry image used as a layer cache.
When using container-features, there are now two images built, so you would need to push two images to the registry and reference two images in the `cacheFrom`.

Making this change will result in a single image build step for the container which will allow for pushing the resulting image to a registry complete with an inline cache manifest, making it possible to get back to the single dev container image that works for image caching.

Anticipated steps:
- [X]  Combine extendImage into build function for Dockerfile #9 
- [X] Update Dockerfile build to use multi-stage build with container-features #13
- [ ] Combine `extendImage` into build function for docker-compose and update docker-compose build to use multi-stage build with container-features (#38)

### PR DESCRIPTION
**Pull request by stuartleeks**
- Add docker-compose tests
- Remove unused imageNameOverride parameter
- docker-compose: combine build and extend
- update docker/docker-compose commands for build/up to detect TTY

This is the final step and closes #10

### CODE REVIEW

File: src/spec-node/devContainersSpecCLI.ts [hunk 1]
```diff
@@ -354,17 +354,19 @@ async function doBuild({
 				throw new Error(`Service '${config.service}' configured in devcontainer.json not found in Docker Compose configuration.`);
 			}
 
-			await buildDockerCompose(config, projectName, buildParams, composeFiles, composeGlobalArgs, [config.service], params.buildNoCache || false, addCacheFroms);
+			const infoOutput = makeLog(buildParams.output, LogLevel.Info);
+			const infoParams = { ...params };
+			infoParams.common.output = infoOutput;
```
[the remark this answers is not in the document]
Comment (stuartleeks) [reply src/spec-node/devContainersSpecCLI.ts:359]: Good spot - I switched to that pattern in other places and meant to come back and clean these up!

[the remark this answers is not in the document]
Comment (stuartleeks) [reply src/spec-node/devContainersSpecCLI.ts]: updated

File: src/spec-node/dockerCompose.ts [hunk 2]
```diff
@@ -107,18 +108,114 @@ export function getRemoteWorkspaceFolder(config: DevContainerFromDockerComposeCo
...
+		// Create overridden Dockerfile and generate docker-compose build override content
+		buildOverrideContent = '    build:\n';
+		const { featureBuildInfo } = extendImageBuildInfo;
+		// We add a '# syntax' line at the start, so strip out any existing line
+		const syntaxMatch = dockerfile.match(/^\s*#\s*syntax\s*=.*[\r\n]/g);
+		if (syntaxMatch) {
+			dockerfile = dockerfile.slice(syntaxMatch[0].length);
+		}
+		let finalDockerfileContent = `${featureBuildInfo.dockerfilePrefixContent}${dockerfile}\n${featureBuildInfo.dockerfileContent}`;
+		const finalDockerfilePath = path.join(featureBuildInfo?.dstFolder, 'Dockerfile-with-features');
```
[the remark this answers is not in the document]
Comment (stuartleeks) [reply src/spec-node/dockerCompose.ts]: done

File: src/spec-node/singleContainer.ts [hunk 3]
```diff
@@ -145,8 +145,8 @@ async function buildAndExtendImage(buildParams: DockerResolverParameters, config
 		if (syntaxMatch) {
 			dockerfile = dockerfile.slice(syntaxMatch[0].length);
 		}
-		let finalDockerfileContent = `${featureBuildInfo.dockerfilePrefixContent}${dockerfile}\n${featureBuildInfo?.dockerfileContent}`;
-		finalDockerfilePath = path.posix.join(featureBuildInfo?.dstFolder, 'Dockerfile-with-features');
+		let finalDockerfileContent = `${featureBuildInfo.dockerfilePrefixContent}${dockerfile}\n${featureBuildInfo.dockerfileContent}`;
+		finalDockerfilePath = path.join(featureBuildInfo?.dstFolder, 'Dockerfile-with-features');
```
[the remark this answers is not in the document]
Comment (stuartleeks) [reply src/spec-node/singleContainer.ts]: done

File: src/spec-node/dockerCompose.ts [hunk 4]
```diff
@@ -209,23 +321,27 @@ async function startContainer(params: DockerResolverParameters, buildParams: Doc
...
 				if (overrideFilePath) {
 					// Add file path to override file as parameter
 					composeGlobalArgs.push('-f', overrideFilePath);
 				}
 			}
 		}
 	} else {
-		await buildDockerCompose(config, projectName, { ...buildParams, output: infoOutput }, localComposeFiles, composeGlobalArgs, config.runServices ?? [], params.buildNoCache ?? false, params.additionalCacheFroms);
-		const { updatedImageName: updatedImageName0, collapsedFeaturesConfig, imageDetails } = await extendImage(params, config, originalImageName, !service.build);
+		const infoParams = { ...params };
```
[the remark this answers is not in the document]
Comment (stuartleeks) [reply src/spec-node/dockerCompose.ts]: done

File: src/spec-node/dockerCompose.ts [hunk 5]
```diff
@@ -107,18 +108,114 @@ export function getRemoteWorkspaceFolder(config: DevContainerFromDockerComposeCo
...
+	const composeConfig = await readDockerComposeConfig(cliParams, localComposeFiles, envFile);
+	const composeService = composeConfig.services[config.service];
+
+	// determine base imageName for generated features build stage(s)
+	let baseName = 'dev_container_auto_added_stage_label';
+	let dockerfile = null;
+	const serviceInfo = getInfoForService(composeService);
+	if (serviceInfo.build) {
+		const { context, dockerfilePath, target } = serviceInfo.build;
+		const resolvedDockerfilePath = path.isAbsolute(dockerfilePath) ? dockerfilePath : path.resolve(context, dockerfilePath);
```
[the remark this answers is not in the document]
Comment (stuartleeks) [reply src/spec-node/dockerCompose.ts]: done

**Comment by stuartleeks** [comment]
Thanks @chrmarti  - have pushed updates for your comments (and rebased into the previous commits to preserve the commit flow for the PR)

**Review by chrmarti — APPROVED** [review]

**Comment by chrmarti** [comment]
Merged, thanks!

File: src/spec-node/dockerCompose.ts [hunk 6]
```diff
@@ -107,19 +108,114 @@ export function getRemoteWorkspaceFolder(config: DevContainerFromDockerComposeCo
...
+			// need to supply a context as we don't have one inherited
+			const emptyDir = cliHost.path.join(await cliHost.tmpdir(), '__devcontainers_cli_empty__');
+			await cliHost.mkdirp(emptyDir);
+			buildOverrideContent += `      context: ${emptyDir}\n`;
+		}
+		// track additional build args to include
+		if (Object.keys(featureBuildInfo.buildArgs).length > 0 || params.buildKitVersion) {
+			buildOverrideContent += '      args:\n';
+			if (params.buildKitVersion) {
+				buildOverrideContent += '        - BUILDKIT_INLINE_CACHE=1\n';
```
Comment (samrocketman) [inline src/spec-node/dockerCompose.ts:196]: `BUILDKIT_INLINE_CACHE=1` breaks end users ability to have a docker-compose file with a `build` Dockerfile step and in that `Dockerfile` referencing an image that is not hosted on DockerHub.

This is an end-user breaking change; when people upgrade this extension (I found out by upgrading).

For example, pre-build my image as `my-custom-image` and then the devcontainer `docker-compose.yml` file references a `Dockerfile` which starts with `FROM my-custom-image`.

The error experienced in the devcontainer build process is:

```
failed to solve: my-custom-image: failed to do request: Head "https://registry-1.docker.io/v2/library/my-custom-image/manifests/latest": blah
```

See also https://github.com/moby/buildkit/issues/2343#issuecomment-1867929392

Comment (stuartleeks) [reply src/spec-node/dockerCompose.ts:196]: @samrocketman  I was reading through the [comment thread](https://github.com/moby/buildkit/issues/2343#issuecomment-1868020509) you linked to - is this still an issue?

**Pull request merged**